### ISSUE TITLE
Moonscript support

### ISSUE THREAD

**Issue by tmandry** [Hammerspoon/hammerspoon#229]
This is more of an idea than a request at this point.

It would be nice to have first-class support for [Moonscript](http://moonscript.org/), a much more expressive language that "compiles" to Lua. In writing a window manager in Lua I personally have found it to be a very dense, inexpressive language, and have found prototypes and (multi-level) metatables very confusing and difficult to deal with. (My WM is fairly complex and has a few different modules, about 1500 lines of Lua altogether so far.) Moonscript can, in theory, fix many of these problems.

Technically, someone who's determined can probably use Moonscript with Hammerspoon now by sticking the moonscript library in `.hammerspoon` and requiring the right things. That said, having the libraries already included would be nice.

#### Error messages
My biggest concern with using things like this (and the reason I haven't switched yet) is error messages.. a traceback to lines of compiled Lua code is not very useful. Moonscript does have a way of dealing with this. Unfortunately, it only does it for you if you use the moonscript cli, not if you're embedding Lua in another program (see https://github.com/leafo/moonscript/issues/106), so we'd have to include some [custom code in HS to get the modified error message traceback](https://github.com/Aegisub/Aegisub/blob/master/libaegisub/lua/utils.cpp#L127).

This has already been done before, so should be pretty straightforward. If we had something like this, we would have first-class support for Moonscript and everyone that uses it will be happy.

What do you think?

**Comment by asmagill** [comment]
If you have luarocks installed via brew, then you can install moonscript from there and anything in the /usr/local tree is available within HS already...  and an adjustment to package.path and/or package.cpath suffices if you install moonscript to a different tree.

Or is there more to embedding it than just supporting require("moonscript")? I've never used it before, but if it's as helpful as you suggest, I'll take a look at it this weekend.

As to including it in the HS release, while I don't have any problems with this in theory, I also don't want to load it down with stuff that's not being used... can you put together some examples of it's use in Hammerspoon? Or at least some comparisons of your code with it and how you'd expect it in moonscript instead?

--
A-Ron

**Comment by mgee** [comment]
At one time my Hammerspoon config was also fully written in Moonscript. However I replaced it with pure Lua code. For me the Moonscript code was much more expressive, but I didn't like the error messages traceback (which can be fixed like it seems).

As an example how a Hammerspoon config in Moonscript can look like in comparison to Lua here is my commit which changed my Hammerspoon Moonscript config to Lua:
https://github.com/mgee/dotfiles/commit/e3e906e7d98d1cd258fcba5fff999982f50b0579

**Comment by cmsj** [comment]
I think my main concern is that it sort of forks the user community, which is still fairly small (we're getting close to 300 people downloading each release, but we're not quite there yet), with some example configs in Lua and some in Moonscript and confusion about which people should use.

I had a quick look at the "custom code" link - is that as part of a re-implementation of Lua in CPP? It didn't look like they had a copy of Lua.

FWIW, we're using Lua from Cocoapods, which means carrying local changes, while not impossible, is something I've tried hard (and so far successfully) to avoid. That's not to say it can never change, but I think keeping a vanilla upstream Lua source tree is a beneficial goal.

I'm rambling a little here, so to sum up. I think overall this would not be a net benefit for the project, however, if the custom traceback support can be added in a way that isn't too invasive, then I would be interested in seeing what it looks like, on the basis that we could integrate the code and then anyone choosing to install moonscript themselves, would be pleasantly delighted to discover that they get useful tracebacks.

It's not something I'm going to be able to spend any time on though, nor would I be able to promise to test it to make sure it doesn't bitrot.

How does that sound?

**Comment by tmandry** [comment]
Yeah, didn't realize we could use luarocks from HS. Bundling it might be premature. But I would like to see the traceback support.

@cmsj, I think the code I pointed to is just a custom error handler function that grabs the moonscript line tables to generate the stack trace (basically a reimplementation of what the moonscript cli does, with handling for regular Lua code as well). A new version of moonscript could break this, but it isn't updated very often (every 6 months or so) and it seems unlikely that this would break in the near future. It could be coded in such a way that it falls back to regular Lua tracebacks and completely non-invasive.

**Comment by jasonm23** [comment]
Anyone can just add moonscript for their own config (very easily), so it's best to just leave it in the hands of the user.

Providing a little FAQ/How to _"How do I use MoonScript for my config?"_ would be a nice touch though.

**Comment by tmandry** [comment]
If I used Moonscript I would want the traceback support, but this isn't important for me right now. I've decided for the time being that it is not worth the added complexity or possibility of splitting the community in two.

However, if I may use this opportunity to rant about the difficulty of hacking Lua, maybe it will inspire some other ideas :) Here are my two main problems:

- Inspecting tables in the console requires an external library! It would be nice if HS included some "magic" to do this automatically (there are potential drawbacks here but I they might be worth it)
- Functional programming is possible, but again, requires external libraries and is messy and not that expressive (who wants to type `hs.fnutils.map(..., function(x) return ... end)` all the time?)

Phew! Okay, that was a good rant. I'm going to close this now, but if someone wants to continue the conversation, feel free.

**Comment by jasonm23** [comment]
I get around this by (temporarily) adding debugging utilities to my config.

I also use Moses to get some decent higher order functions

**Comment by cmsj** [comment]
@tmandry what sort of table inspecting do you want to be doing? we have hs.inspect() that will walk a table and return a string representation.

**Comment by tmandry** [comment]
@jasonm23 Thanks for pointing me to Moses, it looks pretty great.

@cmsj ..yeah, I totally missed that. I was using the `inspect` package from Luarocks. I still think it's annoying from a beginner's perspective that you have to name a function to do something so basic. Chrome/FF consoles are more friendly in this regard.

Like I said in IRC, I'm not hating on Hammerspoon here. I just find that there are some Lua things that have slowed down my dev process that I want to find a solution for. So far this thread has been helpful in that regard :smile:

**Comment by jasonm23** [comment]
By the way, my .hammerspoon is Moonscript.

https://github.com/jasonm23/.hammerspoon

**Comment by asmagill** [comment]
Most places where lua is invoked, the call is handled via lua_xpcall and debug.traceback is the message handler.  If a similar message handler for moonscript could be written (in lua or c), then it would be a simple matter of doing something like

     debug.traceback, oldtraceback = moonscriptTraceback, debug.traceback

in your init.lua file where moonscriptTraceback is the function which replaces debug.traceback.  Keep a copy of oldtraceback around so you can swap it back, if necessary or desired.

It's actually an interesting idea, creating custom traceback mechanisms...  IIRC, Hydra had something not too dissimilar where you could register a custom error handler, but I don't know that it ever went anywhere.  I only came across it in digging through the code when porting stuff over, so I don't know what sdegutis had in mind originally.

**Comment by jasonm23** [comment]
Hi @sdegutis what did you have in mind?

**Comment by sdegutis** [comment]
Hi there. What's the context of this question? Got a tl;dr?

**Comment by jasonm23** [comment]
> Hydra had something not too dissimilar where you could register a custom error handler, but I don't know that it ever went anywhere. I only came across it in digging through the code when porting stuff over, so I don't know what sdegutis had in mind originally.

**Comment by jasonm23** [comment]
Original context is custom trace back handler for moonscript (or anything else in Lua)

**Comment by sdegutis** [comment]
Right, that much I gathered. Still not entirely sure what's being asked and what the goal is.

The custom Lua error handler was meant to capture errors from the user's script so that they could be shown as a dialog box, or so that the user could show it on an alert, or give it a function that takes a string and handle errors however they want. Otherwise errors just went silent and that wasn't helpful to anyone.

Does this help?

**Comment by asmagill** [comment]
k... maybe not as similar as I recalled... I did remember that it was about errors!

However, looking at the moonscript source for the command line (https://github.com/leafo/moonscript/blob/master/bin/moon), all it's doing is:

    ...
      xpcall(run_chunk, function(_err)
        err = _err
        trace = debug.traceback("", 2)
      end)
      if err then
        local truncated = errors.truncate_traceback(util.trim(trace))
        local rewritten = errors.rewrite_traceback(truncated, err)
        if rewritten then
          return print_err(rewritten)
        else
          return print_err(table.concat({
            err,
            util.trim(trace)
          }, "\n"))
        end
    ...

Basically it's a string search & replace process... I don't know much about the internals of moonscript, but it looks like it should be possible to use portions of moonscript itself in a wrapper to debug.traceback.

If I get a chance next week, I may take a look at it, but hopefully this sparks an idea in someone more familiar with it.

**Comment by asmagill** [comment]
Ok.. I don't know moonscript well enough to really test this yet, but for those who want to give it a shot, check out https://github.com/asmagill/hammerspoon-config/blob/master/moonscript_traceback.lua.

Put this file into `~/.hammerspoon/` and you should be able to use the following:

`require("moonscript_traceback").add()`
This will add the moonscript post-processer to debug.traceback

`require("moonscript_traceback").remove()`
This will remove the moonscript post-processer from debug.traceback

As state above, I don't know enough about moonscript to really test this yet... I'll take a look at some of the code referenced above and try it out myself this weekend... but I can confirm that it seems to work properly with regular error tracebacks and gives the same output to the console.

Let me know how it works for you!

**Comment by tmandry** [comment]
Reopening since apparently there is still interest here. @asmagill I haven't actually used Moonscript myself, but perhaps @jasonm23 would try it out?

**Comment by asmagill** [comment]
Sorry @tmandry, saw that you closed this and assumed you were the one interested.  I was mostly interested in knowing how/if we could easily override debug.traceback for our own purposes without having to recompile/change how most of the modules handle call-backs... and I'm happy to say it looks like we can.  Moonscript specifically I leave to those who use it, so unless @jason23 has other questions regarding this, I consider this issue closable again.

**Comment by asmagill** [comment]
@jasonm23 -- Have you had a chance to look at this yet?  I've added it as a gist as well:  https://gist.github.com/asmagill/ffffdb512e7a28398403.  Let me know so we can close this again.

**Comment by jasonm23** [comment]
Looks good, thank you

**Comment by asmagill** [comment]
Closing, unless someone sees a reason to reopen.
